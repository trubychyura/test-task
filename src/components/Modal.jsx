import { connect } from 'react-redux';
import { hideModal } from '../redux/modal';
import { Col, Container, ListGroup, Row } from 'react-bootstrap';
import {
  StyledModal,
  StyledImage,
  StyledDescription,
  StyledComments,
  StyledComentHeader,
} from '../styled/Modal';

import { RatingStars, Form } from '.';
import { useHistory, useParams } from 'react-router';

const Modal = ({ products }) => {
  const params = useParams();
  const history = useHistory();
  const product = products.find((product) => product.asin === params.id);

  if (!product) {
    return null;
  }

  const hideModal = () => {
    history.push('/');
  };

  return (
    <StyledModal show={Boolean(params.id)} onHide={hideModal}>
      <StyledModal.Header closeButton>
        <StyledModal.Title>{product.name}</StyledModal.Title>
      </StyledModal.Header>
      <StyledModal.Body as={Row}>
        <Col lg='auto'>
          <StyledImage src={product.img} />
        </Col>
        <StyledDescription>
          <h5>Description:</h5>
          <p>
            <span>{product.name}</span>
          </p>
        </StyledDescription>
        <StyledComments lg={12}>
          <h5>Comments:</h5>
          <ListGroup>
            {product.comments.map((comment, i) => (
              <ListGroup.Item as={Container} key={i}>
                <StyledComentHeader>
                  <h6>{comment.user || 'Unknown user'}</h6>
                  <RatingStars stars={comment.rating} />
                </StyledComentHeader>
                <p>{comment.text}</p>
              </ListGroup.Item>
            ))}
          </ListGroup>
        </StyledComments>
      </StyledModal.Body>
      <StyledModal.Footer>
        <Form />
      </StyledModal.Footer>
    </StyledModal>
  );
};

export default connect(({ products }) => ({ products }))(Modal);
