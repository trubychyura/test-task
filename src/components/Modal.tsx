import { FC } from 'react';
import { connect } from 'react-redux';
import { Col, Container, ListGroup, Row } from 'react-bootstrap';
import { useHistory, useParams } from 'react-router';

import { RatingStars, Form } from '.';

import { IState, ModalProps, ParamTypes, IProduct } from '../types';
import {
  StyledModal,
  StyledImage,
  StyledDescription,
  StyledComments,
  StyledComentHeader,
} from '../styled/StyledModal';

const Modal: FC<ModalProps> = ({ products }) => {
  const { id } = useParams<ParamTypes>();
  const history = useHistory();
  const product: IProduct | undefined = products.find(
    (product) => product.asin === id,
  );

  if (!product) {
    return null;
  }

  const hideModal = () => {
    history.push('/');
  };

  return (
    <StyledModal show={Boolean(id)} onHide={hideModal}>
      <StyledModal.Header closeButton>
        <StyledModal.Title>{product.name}</StyledModal.Title>
      </StyledModal.Header>
      <StyledModal.Body as={Row}>
        <Col lg='auto'>
          <StyledImage src={product.img} />
        </Col>
        <StyledDescription>
          <h5>Description:</h5>
          <p>
            <span>{product.name}</span>
          </p>
        </StyledDescription>
        <StyledComments lg={12}>
          <h5>Comments:</h5>
          <ListGroup>
            {product.comments.map((comment, i) => (
              <ListGroup.Item as={Container} key={i}>
                <StyledComentHeader>
                  <h6>Unknown user</h6>
                  <RatingStars stars={comment.rating} />
                </StyledComentHeader>
                <p>{comment.text}</p>
              </ListGroup.Item>
            ))}
          </ListGroup>
        </StyledComments>
      </StyledModal.Body>
      <StyledModal.Footer>
        <Form id={id} />
      </StyledModal.Footer>
    </StyledModal>
  );
};

export default connect(({ products }: IState) => ({ products }))(Modal);
