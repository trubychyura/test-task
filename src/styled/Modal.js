import styled from 'styled-components';
import { Col, Form, Image, Modal, Row } from 'react-bootstrap';

export const StyledModal = styled(Modal)`
  @media (max-width: 560px) {
    padding: 0   !important;
  }

  .modal-title {
    text-overflow: ellipsis;
    white-space: nowrap;
    overflow: hidden;
  }

  .modal-dialog {
    max-width: 50%;
    @media (max-width: 1000px) {
      max-width: 70%;
    }
    @media (max-width: 580px) {
      max-width: 90%;
      margin: auto;
    }
  }
`;

export const StyledImage = styled(Image)`
  display: block;
  max-height: 300px;
  max-width: 100%;
  margin: auto;
`;

export const StyledForm = styled(Form)`
  width: 100%;
`;

export const StyledDescription = styled(Col)`
  display: flex;
  flex-direction: column;
  justify-content: center;
`;

export const StyledComments = styled(Col)`
  padding: 15px;
`;

export const StyledComentHeader = styled(Row)`
  margin: 0;
  justify-content: space-between;
`;
