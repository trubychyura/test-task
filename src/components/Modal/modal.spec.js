import { shallow } from 'enzyme';
import routeData from 'react-router';

import { Modal } from './Modal';

import theme from '../../styled/theme';

const products = [
  {
    asin: 'B0721KZHWT',
    name:
      'A3 Magnetic Dry Wipe Weekly Planner Whiteboard for Fridge - Handy Week Calendar/Organiser by Plan Smart - Bonus: 3 Quality Dry Erase Markers with Eraser Included',
    img:
      'https://images-na.ssl-images-amazon.com/images/I/71Y4Ogr8BNL._SX679_.jpg',
    comments: [
      {
        text: 'Some comment',
        rating: 4,
      },
    ],
  },
  {
    asin: 'B01HI4E470',
    name:
      'RoLightic 5M Waterproof Led Strip Light, Multi-Color,SMD 5050 150LEDs Color Changing Flexible Led Strip Kit with 44 key Remote Controller + Power Adapter for Home Kitchen Cabinet Car TV Lighting Decoration',
    img:
      'https://images-na.ssl-images-amazon.com/images/I/81GKt-J0kxL._SX425_.jpg',
    comments: [{ text: 'Some comment', rating: 3 }],
  },
  {
    asin: 'B078PG3PLC',
    name:
      'IPX7 Waterproof Bike Light Set-1000 Lumen Mini Bicycle Headlight Free Rear Light-USB Fast Charge High Performance Rechargeable Battery Easy Install & Universal Fit Mount',
    img:
      'https://images-na.ssl-images-amazon.com/images/I/81iRKp0s5JL._SX522_.jpg',
    comments: [{ text: 'Some comment', rating: 5 }],
  },
];

const setUp = (props) => shallow(<Modal {...props} theme={theme} />);

describe('Form component', () => {
  let component;
  beforeEach(() => {
    jest.spyOn(routeData, 'useParams').mockReturnValue({
      id: 'B0721KZHWT',
    });
    component = setUp({ products });
  });

  it('should render component', () => {
    expect(component).toMatchSnapshot();
  });

  it('should be null', () => {
    expect(setUp({ products: [] }).html()).toBeNull();
  });
});
