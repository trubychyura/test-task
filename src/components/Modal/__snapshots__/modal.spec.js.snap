// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component should render component 1`] = `
<Styled(Modal)
  onHide={[Function]}
  show={true}
>
  <ModalHeader
    closeButton={true}
    closeLabel="Close"
  >
    <ModalTitle>
      A3 Magnetic Dry Wipe Weekly Planner Whiteboard for Fridge - Handy Week Calendar/Organiser by Plan Smart - Bonus: 3 Quality Dry Erase Markers with Eraser Included
    </ModalTitle>
  </ModalHeader>
  <ModalBody
    as={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "defaultProps": Object {
          "noGutters": false,
        },
        "render": [Function],
      }
    }
  >
    <Col
      lg="auto"
    >
      <Styled(Image)
        src="https://images-na.ssl-images-amazon.com/images/I/71Y4Ogr8BNL._SX679_.jpg"
      />
    </Col>
    <Styled(Col)>
      <h5>
        Description:
      </h5>
      <p>
        <span>
          A3 Magnetic Dry Wipe Weekly Planner Whiteboard for Fridge - Handy Week Calendar/Organiser by Plan Smart - Bonus: 3 Quality Dry Erase Markers with Eraser Included
        </span>
      </p>
    </Styled(Col)>
    <Styled(Col)
      lg={12}
    >
      <h5>
        Comments:
      </h5>
      <ListGroup>
        <ListGroupItem
          active={false}
          as={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": Object {
                "fluid": false,
              },
              "render": [Function],
            }
          }
          disabled={false}
          key="0"
        >
          <Styled(Row)>
            <h6>
              Unknown user
            </h6>
            <RatingStars
              stars={4}
            />
          </Styled(Row)>
          <p>
            Some comment
          </p>
        </ListGroupItem>
      </ListGroup>
    </Styled(Col)>
  </ModalBody>
  <ModalFooter>
    <Connect(Form)
      id="B0721KZHWT"
    />
  </ModalFooter>
</Styled(Modal)>
`;
